controller/array_action/array_mediator_svc.py

```diff
@@ -1377,3 +1402,87 @@ def _add_snapshot(self, snapshot_name, source_cli_volume, pool):
+                if NOT_ENOUGH_EXTENTS_IN_POOL_CREATE in ex.my_message:
+                    raise array_errors.NotEnoughSpaceInPool(id_or_name=pool)
+                if any(msg_id in ex.my_message for msg_id in (NAME_NOT_EXIST_OR_MEET_RULES, NON_ASCII_CHARS,
+                                                              INVALID_NAME, TOO_MANY_CHARS)):
+                    raise array_errors.InvalidArgumentError(ex.my_message)
+                raise ex
+        return None
+
+    def _get_cli_volume_id_from_volume_group(self, volume_group_name):
+        filter_value = 'volume_group_name={}'.format(volume_group_name)
+        cli_volume = self._lsvdisk(filtervalue=filter_value)
+        return cli_volume.id
+
+    def _rollback_create_volume_from_snapshot(self, cli_volume_id, volume_group_name):
+        self._rmvolume(cli_volume_id)
+        self._rmvolumegroup(volume_group_name)
+
+    def _change_volume_group(self, cli_volume_id, volume_group):
+        cli_kwargs = {}
+        if volume_group:
+            cli_kwargs['volumegroup'] = volume_group
+        else:
+            cli_kwargs['novolumegroup'] = True
+        self._chvdisk(cli_volume_id, **cli_kwargs)
+
+    def _rename_volume(self, cli_volume_id, name):
+        self._chvdisk(cli_volume_id, name=name)
+
+    def _chvdisk(self, cli_volume_id, **kwargs):
+        try:
+            self.client.svctask.chvdisk(vdisk_id=cli_volume_id, **kwargs)
+        except (svc_errors.CommandExecutionError, CLIFailureError) as ex:
+            if is_warning_message(ex.my_message):
+                logger.warning(
+                    "exception encountered while changing volume parameters '{}': {}".format(kwargs, ex.my_message))
+            else:
+                if OBJ_ALREADY_EXIST in ex.my_message:
+                    raise array_errors.VolumeAlreadyExists(kwargs, self.endpoint)
+                raise ex
+
+    def _rmvolumegroup(self, volume_group_name, not_exist_error=False):
+        logger.info("deleting volume group : {0}".format(volume_group_name))
+        try:
+            self.client.svctask.rmvolumegroup(object_id=volume_group_name)
+        except (svc_errors.CommandExecutionError, CLIFailureError) as ex:
+            if is_warning_message(ex.my_message):
+                logger.warning("exception encountered during deletion of volume group {}: {}".format(volume_group_name,
+                                                                                                     ex.my_message))
+            else:
+                logger.error("Failed to delete volume group {}".format(volume_group_name))
+                if OBJ_NOT_FOUND in ex.my_message or VOL_NOT_FOUND in ex.my_message:
+                    logger.warning(array_errors.ObjectNotFoundError(volume_group_name))
+                    if not not_exist_error:
+                        return
+                raise ex
+
+    def _is_source_has_fcmaps(self, vdisk_uid):
```

Comment:
```suggestion
    def _is_vdisk_has_fcmaps(self, vdisk_uid):
```
this function is not aware that the vdisk is a "source"

Reply:
done